cpp/src/Slice/Parser.cpp

```diff
@@ -5268,72 +5260,76 @@ Slice::Enum::recDependencies(set<ConstructedPtr>&)
     // An Enum does not have any dependencies.
 }
 
-Slice::Enum::Enum(const ContainerPtr& container, const string& name) :
+Slice::Enum::Enum(const ContainerPtr& container, const string& name, const BuiltinPtr& underlying) :
     SyntaxTreeBase(container->unit()),
     Container(container->unit()),
     Type(container->unit()),
     Contained(container, name),
     Constructed(container, name),
+    _underlying(underlying), // can be null
     _explicitValue(false),
-    _minValue(Int32Max),
-    _maxValue(0),
+    _minValue(INT64_MAX),
+    _maxValue(INT64_MIN),
     _lastValue(-1)
 {
 }
 
-int
+int64_t
 Slice::Enum::newEnumerator(const EnumeratorPtr& p)
 {
-    if(p->explicitValue())
+    int64_t rangeMin = 0;
+    uint64_t rangeMax = INT32_MAX;
+    if (_underlying)
     {
-        _explicitValue = true;
-        _lastValue = p->value();
+        tie(rangeMin, rangeMax) = _underlying->integralRange();
+    }
 
-        if(_lastValue < 0)
-        {
-            string msg = "value for enumerator `" + p->name() + "' is out of range";
-            _unit->error(msg);
-        }
+    if (p->explicitValue())
+    {
+        _explicitValue = true;
+        _lastValue  = p->value();
     }
     else
     {
-        if(_lastValue == Int32Max)
+        _lastValue++;
+    }
+
+    if (_lastValue < rangeMin || _lastValue > static_cast<int64_t>(rangeMax))
+    {
+        ostringstream oss;
+        oss.imbue(underscoreSeparatorLocale);
+        oss << "value " << _lastValue << " for enumerator `" << p->name() << "'";
+        if (_underlying)
         {
-            string msg = "value for enumerator `" + p->name() + "' is out of range";
-            _unit->error(msg);
+            oss << " is outside type's " << _underlying->kindAsString() << " range: [" << rangeMin << ".." << rangeMax
+                << "]";
         }
         else
         {
-            //
-            // If the enumerator was not assigned an explicit value, we automatically assign
-            // it one more than the previous enumerator.
-            //
-            ++_lastValue;
+            oss << " is out of range";
```

Comment:
In the case where no explicit underlying type is declared, the range should default to varint, correct?
I think it would still be useful to print out the range so the user knows what values are allowed instead of having to guess or look it up in the docs.

Reply:
It's actually 0..INT32_MAX, which does not correspond to a builtin type.

Reply:
Even more reason to include the actual limits in the error message then, otherwise the user is just left to guess the limits.

Reply:
Fixed.